File: python/examples/ilinks_api_example.py
"""Example demonstrating the ILinks flat API"""

from pathlib import Path
from links_client.api.ilinks import ILinks, LinkConstants


def main():
    print("=== ILinks Flat API Example ===\n")

    # Create ILinks instance
    db_path = Path("data/example-ilinks.links")
    links = ILinks(str(db_path))

    # Get constants
    constants = links.get_constants()
    print(f"Constants: Continue={constants.CONTINUE}, Break={constants.BREAK}, Any={constants.ANY.value}")

    # Create some links
    print("\n--- Creating Links ---")
    link1_id = links.create([1, 2])
    print(f"Created link (1 -> 2) with ID: {link1_id}")

    link2_id = links.create([3, 4])
    print(f"Created link (3 -> 4) with ID: {link2_id}")

    link3_id = links.create([1, 5])
    print(f"Created link (1 -> 5) with ID: {link3_id}")

    # Count links
    print("\n--- Counting Links ---")
    total_count = links.count()
    print(f"Total links in database: {total_count}")

    count_with_source_1 = links.count([1, constants.ANY.value])
    print(f"Links with source=1: {count_with_source_1}")

    # Iterate through links
    print("\n--- Iterating Through Links ---")
    iteration_count = 0

    def iterator(link):
        nonlocal iteration_count
        print(f"Link {link['id']}: {link['source']} -> {link['target']}")
        iteration_count += 1
        if iteration_count >= 3:
            return constants.BREAK
        return constants.CONTINUE

    links.each(None, iterator)

    # Update a link
    print("\n--- Updating Link ---")
    print(f"Updating link {link1_id} to (10 -> 20)")

    def update_handler(change):
        print("Before:", change["before"])
        print("After:", change["after"])

    links.update([link1_id, constants.ANY.value, constants.ANY.value], [10, 20], update_handler)

    # Read updated link
    updated_links = []

    def read_updated(link):
        updated_links.append(link)
        return constants.CONTINUE

    links.each([link1_id, constants.ANY.value, constants.ANY.value], read_updated)
    print("Updated link:", updated_links[0])

    # Delete a link
    print("\n--- Deleting Link ---")
    print(f"Deleting link {link2_id}")

    def delete_handler(change):
        print("Deleted:", change["before"])

    links.delete([link2_id, constants.ANY.value, constants.ANY.value], delete_handler)

    # Count after deletion
    final_count = links.count()
    print(f"\nFinal link count: {final_count}")

    print("\n=== Example Complete ===")


if __name__ == "__main__":
    main()

File: js/examples/ilinks-api-example.js
// Example demonstrating the ILinks flat API
import ILinks from '../src/api/ilinks.js';
import path from 'path';

async function main() {
  console.log('=== ILinks Flat API Example ===\n');

  // Create ILinks instance
  const dbPath = path.join(process.cwd(), 'data', 'example-ilinks.links');
  const links = new ILinks(dbPath);

  // Get constants
  const constants = links.getConstants();
  console.log('Constants:', { Continue: constants.Continue, Break: constants.Break, Any: constants.Any });

  // Create some links
  console.log('\n--- Creating Links ---');
  const link1Id = await links.create([1, 2]);
  console.log(`Created link (1 -> 2) with ID: ${link1Id}`);

  const link2Id = await links.create([3, 4]);
  console.log(`Created link (3 -> 4) with ID: ${link2Id}`);

  const link3Id = await links.create([1, 5]);
  console.log(`Created link (1 -> 5) with ID: ${link3Id}`);

  // Count links
  console.log('\n--- Counting Links ---');
  const totalCount = await links.count();
  console.log(`Total links in database: ${totalCount}`);

  const countWithSource1 = await links.count([1, constants.Any]);
  console.log(`Links with source=1: ${countWithSource1}`);

  // Iterate through links
  console.log('\n--- Iterating Through Links ---');
  let iterationCount = 0;
  await links.each(null, (link) => {
    console.log(`Link ${link.id}: ${link.source} -> ${link.target}`);
    iterationCount++;
    if (iterationCount >= 3) {
      return constants.Break; // Stop after 3 links
    }
    return constants.Continue;
  });

  // Update a link
  console.log('\n--- Updating Link ---');
  console.log(`Updating link ${link1Id} to (10 -> 20)`);
  await links.update([link1Id, constants.Any, constants.Any], [10, 20], (change) => {
    console.log('Before:', change.before);
    console.log('After:', change.after);
  });

  // Read updated link
  const updatedLinks = [];
  await links.each([link1Id, constants.Any, constants.Any], (link) => {
    updatedLinks.push(link);
    return constants.Continue;
  });
  console.log('Updated link:', updatedLinks[0]);

  // Delete a link
  console.log('\n--- Deleting Link ---');
  console.log(`Deleting link ${link2Id}`);
  await links.delete([link2Id, constants.Any, constants.Any], (change) => {
    console.log('Deleted:', change.before);
  });

  // Count after deletion
  const finalCount = await links.count();
  console.log(`\nFinal link count: ${finalCount}`);

  console.log('\n=== Example Complete ===');
}

main().catch(console.error);
